format-sdk/src/main/cpp/convert/column_converter.cpp

```diff
@@ -0,0 +1,228 @@
+            auto sr_key_field = std::make_shared<starrocks::Field>(sr_field->sub_field(0));
+            auto sr_value_field = std::make_shared<starrocks::Field>(sr_field->sub_field(1));
+            ARROW_ASSIGN_OR_RAISE(auto converter,
+                                  ColumnConverter::create(arrow_offset_datatype, sr_offset_field, pool));
+            result.emplace_back(converter);
+            ARROW_ASSIGN_OR_RAISE(converter, ColumnConverter::create(key_type, sr_key_field, pool));
+            result.emplace_back(converter);
+            ARROW_ASSIGN_OR_RAISE(converter, ColumnConverter::create(value_type, sr_value_field, pool));
+            result.emplace_back(converter);
+        } else if (arrow_type->id() == arrow::Type::STRUCT &&
+                   sr_field->type()->type() == starrocks::LogicalType::TYPE_STRUCT) {
+            if (!sr_field->has_sub_fields()) {
+                return arrow::Status::Invalid("Has no sub fields for struct field ", sr_field->name());
+            }
+            if (arrow_type->num_fields() != sr_field->sub_fields().size()) {
+                return arrow::Status::Invalid("Field number mismatch, arrow: ", arrow_type->num_fields(),
+                                              ", sr: ", sr_field->sub_fields().size());
+            }
+
+            for (size_t idx = 0; idx < arrow_type->num_fields(); ++idx) {
+                if (arrow_type->field(idx)->name() != sr_field->sub_field(idx).name()) {
+                    return arrow::Status::Invalid("Field name mismatch, arrow: ", arrow_type->field(idx)->name(),
+                                                  ", sr: ", sr_field->sub_field(idx).name());
+                }
+                std::shared_ptr<starrocks::Field> sr_child_field =
+                        std::make_shared<starrocks::Field>(sr_field->sub_field(idx));
+
+                ARROW_ASSIGN_OR_RAISE(auto converter,
+                                      ColumnConverter::create(arrow_type->field(idx)->type(), sr_child_field, pool));
+                result.emplace_back(converter);
+            }
+        } else {
+            return arrow::Status::NotImplemented(
+                    "Unsupported conversion between ", arrow_type->ToString(), " and ", sr_field->to_string());
+        }
+        return result;
+    }
+
+    /*
+     * ColumnConverter member functions
+     */
+
+    /*
+     * Convert starrocks null data to arrow null bitmap:
+     * - arrow use 0 as null indicator and use bit
+     * - while sr use 1 as null indicator and use byte
+     */
+    arrow::Result<std::shared_ptr<arrow::Buffer>> ColumnConverter::convert_null_bitmap(
+            const Buffer<uint8_t> &null_bytes) {
+        std::shared_ptr<arrow::Buffer> null_bitmap;
+        ARROW_ASSIGN_OR_RAISE(
+                null_bitmap, arrow::internal::BytesToBits(
+                reinterpret_cast<const std::vector<uint8_t> &>(null_bytes),
+                const_cast<arrow::MemoryPool *>(_pool))
```

Comment:
To make `_pool` as const is mainly to prevent it from being changed in the column converter, but the param `pool` received by the `BytesToBits` method is a non-const type, so a `const_cast` conversion is done as a last resort.